### ISSUE TITLE
expo router config plugin: `Error: Cannot find module 'ajv/dist/compile/codegen'`

### ISSUE THREAD

**Issue by BalogunofAfrica** [expo/expo#26641]
### Minimal reproducible example

https://gist.github.com/BalogunofAfrica/50b3dd70c47c56ccdda10594ad71518b

### Which package manager are you using? (Yarn is recommended)

yarn

### If the issue is web-related, please select the bundler (`web.bundler` in the `app.json`)

None

### Summary

NOTE: This issue only happens when you use `bun install`, with `yarn` it works fine. (`bun` is not an option in package manager option above)

Running `bunx expo start` fails with the error:

```sh
Error: Cannot find module 'ajv/dist/compile/codegen'
Require stack:
- /path-to-project/node_modules/ajv-keywords/dist/definitions/typeof.js
- /path-to-project/node_modules/ajv-keywords/dist/keywords/typeof.js
- /path-to-project/node_modules/ajv-keywords/dist/keywords/index.js
- /path-to-project/node_modules/ajv-keywords/dist/index.js
- /path-to-project/node_modules/schema-utils/dist/validate.js
- /path-to-project/node_modules/schema-utils/dist/index.js
- /path-to-project/node_modules/expo-router/plugin/build/index.js
- /path-to-project/node_modules/expo-router/app.plugin.js
- /path-to-project/node_modules/@expo/config-plugins/build/utils/plugin-resolver.js
- /path-to-project/node_modules/@expo/config-plugins/build/plugins/withStaticPlugin.js
- /path-to-project/node_modules/@expo/config-plugins/build/plugins/withPlugins.js
- /path-to-project/node_modules/@expo/config-plugins/build/index.js
- /path-to-project/node_modules/@expo/cli/build/src/utils/scheme.js
- /path-to-project/node_modules/@expo/cli/build/src/start/resolveOptions.js
- /path-to-project/node_modules/@expo/cli/build/src/start/index.js
- /path-to-project/node_modules/@expo/cli/build/bin/cli
- /path-to-project/node_modules/expo/bin/cli
Error: Cannot find module 'ajv/dist/compile/codegen'
Require stack:
- /path-to-project/node_modules/ajv-keywords/dist/definitions/typeof.js
- /path-to-project/node_modules/ajv-keywords/dist/keywords/typeof.js
- /path-to-project/node_modules/ajv-keywords/dist/keywords/index.js
- /path-to-project/node_modules/ajv-keywords/dist/index.js
- /path-to-project/node_modules/schema-utils/dist/validate.js
- /path-to-project/node_modules/schema-utils/dist/index.js
- /path-to-project/node_modules/expo-router/plugin/build/index.js
- /path-to-project/node_modules/expo-router/app.plugin.js
- /path-to-project/node_modules/@expo/config-plugins/build/utils/plugin-resolver.js
- /path-to-project/node_modules/@expo/config-plugins/build/plugins/withStaticPlugin.js
- /path-to-project/node_modules/@expo/config-plugins/build/plugins/withPlugins.js
- /path-to-project/node_modules/@expo/config-plugins/build/index.js
- /path-to-project/node_modules/@expo/cli/build/src/utils/scheme.js
- /path-to-project/node_modules/@expo/cli/build/src/start/resolveOptions.js
- /path-to-project/node_modules/@expo/cli/build/src/start/index.js
- /path-to-project/node_modules/@expo/cli/build/bin/cli
- /path-to-project/node_modules/expo/bin/cli
    at Module._resolveFilename (node:internal/modules/cjs/loader:1134:15)
    at Module._load (node:internal/modules/cjs/loader:975:27)
    at Module.require (node:internal/modules/cjs/loader:1225:19)
    at require (node:internal/modules/helpers:177:18)
    at Object.<anonymous> (/path-to-project/node_modules/ajv-keywords/dist/definitions/typeof.js:3:19)
    at Module._compile (node:internal/modules/cjs/loader:1356:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1414:10)
    at Module.load (node:internal/modules/cjs/loader:1197:32)
    at Module._load (node:internal/modules/cjs/loader:1013:12)
    at Module.require (node:internal/modules/cjs/loader:1225:19)
```

package info:
```json
{
    "expo": "^49.0.21",
    "expo-router": "2.0.0"
}
```

### Environment

```sh
  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.2.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 18.19.0 - ~/.nvm/versions/node/v18.19.0/bin/node
      Yarn: 1.22.21 - ~/.bun/bin/yarn
      npm: 10.2.3 - ~/.nvm/versions/node/v18.19.0/bin/npm
      Watchman: 2023.12.04.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.12.1 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.2, iOS 17.2, macOS 14.2, tvOS 17.2, watchOS 10.2
    IDEs:
      Android Studio: 2022.3 AI-223.8836.35.2231.10811636
      Xcode: 15.1/15C65 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^49.0.21 => 49.0.21 
      expo-router: 2.0.0 => 2.0.0 
      react: 18.2.0 => 18.2.0 
      react-dom: 18.2.0 => 18.2.0 
      react-native: 0.72.6 => 0.72.6 
    Expo Workflow: managed
```

**Comment by Cheesetouched** [comment]
Facing this as well. I am also using bun. `bun install -d ajv` fixed it. Is there a recommended workaround though?

**Comment by NoZiL** [comment]
I am also facing this issue, due to `eslint` using a different `ajv` version that may be conflicting with the one resolved by `bun` for `eslint`.

Here is a minimal reproducible example:
```bash
bunx create-expo-app@latest --template tabs@50 with-router # from https://docs.expo.dev/router/installation/#quick-start
cd with-router
bun add eslint prettier eslint-config-universe --save-dev # from https://docs.expo.dev/guides/using-eslint/
bunx eas-cli build --platform android --local
```

The last command would fail with `Cannot find module 'ajv/dist/compile/codegen'`

Hints: `ajv` is not referenced as a dependency in `expo-router` package.json. Maybe using the plugin would require to pass it as a ` peerDependency` for `bun` to automatically resolve it.

**Comment by sammoore** [comment]
I looked into this, for us (and the `create-expo-stack` template), it seems to be because `eslint` pulls in an older version of `ajv@6.12.4`, and bun incorrectly de-dupes it with the one `ajv-keywords` pulls in (via `schema-utils` via `expo-router`).

![Screen Shot 2024-02-10 at 5 18 59 PM](https://github.com/expo/expo/assets/2035492/f84db5dc-8d35-4c56-80e1-a1c169f6183c)

I'm going to report this to bun. The workaround posted above works for now:

![Screen Shot 2024-02-10 at 5 18 48 PM](https://github.com/expo/expo/assets/2035492/66188d12-3bd9-4ca1-9cc4-9785d9a5349d)

**Comment by c0d3x** [comment]
After I added the ajv plugin I now get linting errors:

```SH
bun run lint:fix
$ eslint . --ext .ts,.tsx,.js,.jsx --fix

NOT SUPPORTED: option missingRefs. Pass empty schema with $id that should be ignored to ajv.addSchema.

Oops! Something went wrong! :(

ESLint: 8.57.0

TypeError: Cannot set properties of undefined (setting 'defaultMeta')
    at ajvOrig (/Users/x/projects/expo-test/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:1626:27)
    at Object.<anonymous> (/Users/x/projects/expo-test/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:1924:13)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1435:10)
    at Module.load (node:internal/modules/cjs/loader:1207:32)
    at Module._load (node:internal/modules/cjs/loader:1023:12)
    at Module.require (node:internal/modules/cjs/loader:1235:19)
    at require (node:internal/modules/helpers:176:18)
    at Object.<anonymous> (/Users/x/projects/expo-test/node_modules/eslint/lib/cli-engine/cli-engine.js:33:5)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
```

**Comment by isaachinman** [comment]
@c0d3x I saw that error as well. I pinned bun to v1.0.29. The bun team seems to be dropping the ball recently. I noticed that my actual lockfile changed between v1.0.29 and v1.0.30, for example.

No idea what is going on, but it is not conducive to writing stable applications.

As far as I am concerned, bun can only be seriously considered if you pin your version and upgrade intentionally/carefully.
